channeldb/migration27/migration.go

```diff
@@ -0,0 +1,118 @@
+package migration27
+
+import (
+	"bytes"
+	"fmt"
+
+	mig "github.com/lightningnetwork/lnd/channeldb/migration_01_to_11"
+	"github.com/lightningnetwork/lnd/kvdb"
+)
+
+var (
+	// historicalChannelBucket stores all channels that have seen their
+	// commitment tx confirm. All information from their previous open state
+	// is retained.
+	historicalChannelBucket = []byte("historical-chan-bucket")
+)
+
+// MigrateHistoricalBalances patches the two new fields, `InitialLocalBalance`
+// and `InitialRemoteBalance`, for all the open channels saved in historical
+// channel bucket. Unlike migration 25, it will only read the old channel info
+// first and then patch the new tlv records with empty values. For historical
+// channels, we previously didn't save the initial balances anywhere and since
+// it's corresponding open channel bucket is deleted after closure, we have
+// lost that balance info.
+func MigrateHistoricalBalances(tx kvdb.RwTx) error {
+	log.Infof("Migrating historical local and remote balances...")
+
+	// First fetch the top level bucket which stores all data related to
+	// historically stored channels.
+	rootBucket := tx.ReadWriteBucket(historicalChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if rootBucket == nil {
+		return nil
+	}
+
+	// Read a list of historical channels.
+	channels, err := findHistoricalChannels(rootBucket)
```

Comment:
IIRC the bucket has been around for 1.5 years or so, but _does_ include every channel ever opened by the node. So it can be large, but likely doesn't have _millions_ of entries.

Reply:
Ok cool. Each `ChannelInfo` is below 1KB (roughly 913 bytes if funding tx is 250 bytes), a 1 GB ram should fit 1 million records. As long as we don't have millions of channels here we'd be fine.